### ISSUE TITLE
Colored mesh

### ISSUE THREAD

**Issue by PavelBlend** [doyubkim/fluid-engine-dev#238]
Hello.

I have particles that have color:
![01](https://user-images.githubusercontent.com/7983249/54545081-fd58b900-49b1-11e9-8e18-3a09ccf19feb.jpg)
Is it possible to transfer these colors to the mesh of a liquid?

**Comment by doyubkim** [comment]
This feature is not available at the moment but sounds like a good addition. Will add to the feature list!

**Comment by PavelBlend** [comment]
It turned out to blender means to implement the painting of the mesh:
![rend_colored_mesh](https://user-images.githubusercontent.com/7983249/55098636-2f170180-50cf-11e9-892f-1d99040abf45.jpg)

**Comment by doyubkim** [comment]
Thanks for the experiment! Yes, I think that’s a good idea. Let me try letting particles to sample texture coordinates from an input mesh and carry the values throughout simulations.

**Comment by kentbarber** [comment]
I would also like to be able to set the color of particles and also grid cells. Have the colors of cells advected throughout the domain. I am happy to set the colours myself, I don't need any sampling from textures done by the framework itself. It would also be nice to set the color for the emitted particles as a parameter to the emitters, just set a Vector3D(red, green, blue) between 1-0.

**Comment by utilForever** [comment]
I'll take a look this weekend.

**Comment by utilForever** [comment]
Are you want to add feature for setting the color of particles and grid cells?

**Comment by kentbarber** [comment]
Amazing! Thank you. Particles covers almost everything except LevelSet. So I thought it would be good to have a solution for that as well, which is why I mentioned being able to set a color for the grid cells as well. I could then extract out the color from the cell and use that to apply vertex colours to the generated mesh myself, as well as allowing me to possibly blend neighbouring cell colours to get a smoother interpolation. So both would be great, but if you can only get the particle colour that would be great anyway. Thanks a lot, really appreciate the work you and Doyub are doing here.

**Comment by kentbarber** [comment]
Hi @utilForever, just wondering if you have added a color support for particles yet? If not then I may start on it myself. I won't be worry about LevelSetSolver for now and will just add support throughout for setting a color for particle emitters and having that data stored for each individual particle in the ParticleSystemData.

**Comment by kentbarber** [comment]
@utilForever I decided to start with this. Looks simple enough to add. So if your not working on this yet then do let me know, otherwise I could submit my own changes for a pull request.

**Comment by kentbarber** [comment]
Actually I think what I am doing is not actually part of this particular Issue. Since what I am doing is adding colors directly to the Jet Framework and this issue is about using particle colors to set the vertex colors of a mesh, these are two different topics although they are related. I am going to open a new issue for this.

**Comment by doyubkim** [comment]
@kentbarber @utilForever Here's what I was envisioning for this feature:

For the particle-based solvers, add additional channel (color channel) to `ParticleSystemData3`. After performing mesh generation using marching cubes, enumerate over the mesh vertices, "interpolate" colors around the vertices using SPH-style interapolation and/or nearest neighbor approach (maybe have an option to pick one). Then the colored mesh could be achieved.

For the grid-based solvers, add color channel to `GridSystemData3`, make them "advectable", and when performing marching cubes, directly interpolate color values from the grid which now has color channel.

@kentbarber if you are already working on the particles, then please feel free to post a PR! We can discuss from there.

**Comment by utilForever** [comment]
@kentbarber Oh, I'm sorry. I've been very busy with my company.
Are you started implementation? If you not working on this yet, I'd like to start next week.

**Comment by kentbarber** [comment]
@utilForever Yes I have started adding support for particles, but it is still under discussion. I have opened a new issue for adding color information to particles: https://github.com/doyubkim/fluid-engine-dev/issues/277#issuecomment-562773481

I haven't touched GridSystemData3 yet, so if you would like to do that then feel free. I would consider it to be a separate task and perhaps could be its own issue as well. 

Then once color has been added to both of the above, this particular issue could be used to track the interpolation task for creating the color for the generated mesh.

**Comment by utilForever** [comment]
@kentbarber Hmm... If you haven't started implementing the code yet, I'd like to work on both.
I think that since color data is not essential, it would be good to provide additional data as an extension such as `std::optional` or new design.

**Comment by kentbarber** [comment]
@utilForever That is ok with me. I will hold off on any further work in the Jet Framework itself and will instead get my code ready to create and use the color data once it is in place. I would prefer the ability to add arbitrary data to the system, with color being one of them, so a more general case system that is usable throughout Jet would be a better solution. Thanks.

**Comment by utilForever** [comment]
@kentbarber Thank you for your consideration.
I'll discuss this design on #277.
